subprojects/core/src/main/java/org/gradle/internal/buildtree/BuildTreeScopeServices.java

```diff
@@ -69,4 +78,35 @@ protected ExceptionAnalyser createExceptionAnalyser(ListenerManager listenerMana
     public DefaultProjectStateRegistry createProjectPathRegistry(WorkerLeaseService workerLeaseService) {
         return new DefaultProjectStateRegistry(workerLeaseService);
     }
+
+    ValidateStep.ValidationWarningReporter createValidationWarningReporter() {
+        return new WorkValidationWarningReporterImpl();
+    }
+
+    private static class WorkValidationWarningReporterImpl implements ValidateStep.ValidationWarningReporter, Closeable {
```

Comment:
```suggestion
    private static class DefaultWorkValidationWarningReporter implements ValidateStep.ValidationWarningReporter, Closeable {
```

How about using the `Default` prefix as we normally do for this class instead of an `Impl` postfix. It also seems like a good idea to move this implementation to its own file. WDYT?

Reply:
Good point about the type name, renamed. I'd rather keep it here, though.

Reply:
Actually, I convinced myself otherwise, and moved it to a top-level type.